File: src/color_ops.py
import cv2
import numpy as np

class ColorOps:
    @staticmethod
    def gray(frames):
        result = []
        for f in frames:
            g = cv2.cvtColor(f, cv2.COLOR_BGR2GRAY)
            result.append(cv2.cvtColor(g, cv2.COLOR_GRAY2BGR))
        return result
    
    @staticmethod
    def invert(frames):
        return [cv2.bitwise_not(f) for f in frames]
    
    @staticmethod
    def brightness(frames, value):
        return [cv2.convertScaleAbs(f, alpha=1, beta=value) for f in frames]
    
    @staticmethod
    def contrast(frames, value):
        return [cv2.convertScaleAbs(f, alpha=value, beta=0) for f in frames]
    
    @staticmethod
    def hue_shift(frames, shift):
        result = []
        for f in frames:
            hsv = cv2.cvtColor(f, cv2.COLOR_BGR2HSV).astype(np.float32)
            hsv[:,:,0] = (hsv[:,:,0] + shift) % 180
            result.append(cv2.cvtColor(hsv.astype(np.uint8), cv2.COLOR_HSV2BGR))
        return result
    
    @staticmethod
    def saturation(frames, factor):
        result = []
        for f in frames:
            hsv = cv2.cvtColor(f, cv2.COLOR_BGR2HSV).astype(np.float32)
            hsv[:,:,1] = np.clip(hsv[:,:,1] * factor, 0, 255)
            result.append(cv2.cvtColor(hsv.astype(np.uint8), cv2.COLOR_HSV2BGR))
        return result
    
    @staticmethod
    def sepia(frames):
        result = []
        sepia_matrix = np.array([[0.272, 0.534, 0.131],
                                [0.349, 0.686, 0.168],
                                [0.393, 0.769, 0.189]])
        for f in frames:
            sepia_frame = cv2.transform(f, sepia_matrix)
            result.append(np.clip(sepia_frame, 0, 255).astype(np.uint8))
        return result
    
    @staticmethod
    def warm(frames, intensity):
        result = []
        for f in frames:
            b, g, r = cv2.split(f.astype(np.float32))
            r = np.clip(r + intensity, 0, 255)
            b = np.clip(b - intensity * 0.5, 0, 255)
            result.append(cv2.merge([b, g, r]).astype(np.uint8))
        return result
    
    @staticmethod
    def cool(frames, intensity):
        result = []
        for f in frames:
            b, g, r = cv2.split(f.astype(np.float32))
            b = np.clip(b + intensity, 0, 255)
            r = np.clip(r - intensity * 0.5, 0, 255)
            result.append(cv2.merge([b, g, r]).astype(np.uint8))
        return result
    
    @staticmethod
    def posterize(frames, levels):
        result = []
        step = 256 // levels
        for f in frames:
            quantized = (f // step) * step
            result.append(quantized.astype(np.uint8))
        return result

File: src/ColorOps.cpp
#include "../include/ColorOps.h"

using namespace cv;

std::vector<Mat> ColorOps::gray(const std::vector<Mat>& frames) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        Mat g, bgr;
        cvtColor(f, g, COLOR_BGR2GRAY);
        cvtColor(g, bgr, COLOR_GRAY2BGR);
        result.push_back(bgr);
    }
    return result;
}

std::vector<Mat> ColorOps::invert(const std::vector<Mat>& frames) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        Mat inv; bitwise_not(f, inv); result.push_back(inv);
    }
    return result;
}

std::vector<Mat> ColorOps::brightness(const std::vector<Mat>& frames, int value) {
    std::vector<Mat> result;
    for (const auto& f : frames) { Mat bright; f.convertTo(bright, -1, 1, value); result.push_back(bright); }
    return result;
}

std::vector<Mat> ColorOps::contrast(const std::vector<Mat>& frames, double value) {
    std::vector<Mat> result;
    for (const auto& f : frames) { Mat cont; f.convertTo(cont, -1, value, 0); result.push_back(cont); }
    return result;
}

std::vector<Mat> ColorOps::hueShift(const std::vector<Mat>& frames, int shift) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        Mat hsv; cvtColor(f, hsv, COLOR_BGR2HSV); std::vector<Mat> channels; split(hsv, channels);
        channels[0].convertTo(channels[0], CV_32F); channels[0] = channels[0] + shift;
        Mat temp; channels[0].copyTo(temp); temp = temp / 180.0; temp.forEach<float>([](float &p, const int*) { p = std::floor(p); }); temp = temp * 180.0;
        channels[0] = cv::max(channels[0] - temp, 0.0f);
        channels[0].convertTo(channels[0], CV_8U); Mat merged; merge(channels, merged); Mat bgr; cvtColor(merged, bgr, COLOR_HSV2BGR); result.push_back(bgr);
    }
    return result;
}

std::vector<Mat> ColorOps::saturation(const std::vector<Mat>& frames, double factor) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        Mat hsv; cvtColor(f, hsv, COLOR_BGR2HSV); std::vector<Mat> channels; split(hsv, channels);
        channels[1].convertTo(channels[1], CV_32F); channels[1] = channels[1] * factor; channels[1] = min(channels[1], 255.0); channels[1] = max(channels[1], 0.0); channels[1].convertTo(channels[1], CV_8U);
        Mat merged; merge(channels, merged); Mat bgr; cvtColor(merged, bgr, COLOR_HSV2BGR); result.push_back(bgr);
    }
    return result;
}

std::vector<Mat> ColorOps::sepia(const std::vector<Mat>& frames) {
    std::vector<Mat> result; Mat sepiaMatrix = (Mat_<float>(3,3) << 0.272f,0.534f,0.131f, 0.349f,0.686f,0.168f, 0.393f,0.769f,0.189f);
    for (const auto& f : frames) { Mat sepiaFrame; transform(f, sepiaFrame, sepiaMatrix); sepiaFrame = min(sepiaFrame, 255.0); sepiaFrame = max(sepiaFrame, 0.0); sepiaFrame.convertTo(sepiaFrame, CV_8U); result.push_back(sepiaFrame); }
    return result;
}

std::vector<Mat> ColorOps::warm(const std::vector<Mat>& frames, int intensity) {
    std::vector<Mat> result; for (const auto& f : frames) { std::vector<Mat> channels; split(f, channels); channels[0].convertTo(channels[0], CV_32F); channels[1].convertTo(channels[1], CV_32F); channels[2].convertTo(channels[2], CV_32F); channels[2] = min(channels[2] + intensity, 255.0); channels[0] = max(channels[0] - intensity * 0.5, 0.0); channels[0].convertTo(channels[0], CV_8U); channels[1].convertTo(channels[1], CV_8U); channels[2].convertTo(channels[2], CV_8U); Mat merged; merge(channels, merged); result.push_back(merged); }
    return result;
}

std::vector<Mat> ColorOps::cool(const std::vector<Mat>& frames, int intensity) {
    std::vector<Mat> result; for (const auto& f : frames) { std::vector<Mat> channels; split(f, channels); channels[0].convertTo(channels[0], CV_32F); channels[1].convertTo(channels[1], CV_32F); channels[2].convertTo(channels[2], CV_32F); channels[0] = min(channels[0] + intensity, 255.0); channels[2] = max(channels[2] - intensity * 0.5, 0.0); channels[0].convertTo(channels[0], CV_8U); channels[1].convertTo(channels[1], CV_8U); channels[2].convertTo(channels[2], CV_8U); Mat merged; merge(channels, merged); result.push_back(merged); }
    return result;
}

std::vector<Mat> ColorOps::posterize(const std::vector<Mat>& frames, int levels) {
    std::vector<Mat> result; int step = 256 / levels; for (const auto& f : frames) { Mat quantized; f.convertTo(quantized, CV_32F); quantized = (quantized / step) * step; quantized = min(quantized, 255.0); quantized = max(quantized, 0.0); quantized.convertTo(quantized, CV_8U); result.push_back(quantized); } return result;
}
